### ISSUE TITLE
torch.where leads to unexpected gradients

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#52248]
```py
In [58]: a = torch.randn(()).requires_grad_()

In [59]: b = torch.tensor(False)

In [60]: c = torch.ones(())

In [61]: torch.where(b, a/0, c)
Out[61]: tensor(1., grad_fn=<SWhereBackward>)

In [62]: torch.autograd.grad(torch.where(b, a/0, c), a)
Out[62]: (tensor(nan),)
```

This is yet another case of weird behavior because we don't differentiate "no gradient" vs. "zero gradient". The not-used branch computes gradient 0/0 and that becomes NaN... 

This makes writing NaN avoiding ops kinda difficult without doing a custom autograd Function. 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/50122

**Comment by albanD** [comment]
I don't think there is anything we can do here...
Or at least until we introduce a ZeroTensor for zeros and IndependentGradTensor for independent.
But then user code will need to use the right one as well (in custom Functions in particular).

**Comment by SsnL** [comment]
Well.... @albanD  is there way to shorthand (maybe via jit) transform
```py
torch.where(pred, f(x), g(x))
```
to
```py
class Fn(Function):
  # forward is what above is
  def backward(ctx, gout):
    gx = torch.empty_like(gout)
    gx[pred] = BWD_FORMULA_FOR_f(x[pred], gout[pred])
    gx[~pred] = BWD_FORMULA_FOR_g(x[~pred], gout[~pred])

Fn.apply(x)
```
ideally supporting multiple args.

**Comment by albanD** [comment]
What you mean here is giving the `/` as input to this custom function as well?
Indeed collapsing the `/` and `where` into a single elementary op allows you to solve that problem, but you need to write  a backward for that...

The best workaround is to avoid the `/0` altogether, usually by moving it just after the where op.

**Comment by jatentaki** [comment]
@SsnL could you elaborate on how to implement your workaround as a one-off patch (i.e. I don't need it as an automatic `jit` transform)? More specifically, how can I elicit the `BWD_FORMULA_FOR_*` if my `f, g` are rather involved?

My use-case is similar to semantic segmentation, where at some locations I am missing the proper ground truth so I use a weaker supervision signal. I use `NaN` to indicate regions where the strong GT is unavailable and then use
```
torch.where(torch.isfinite(strong_supervision(x, labels)), strong_supervision(x, labels), weak_supervision(x))
```
 to merge the results. Since for a long time I did not need to backprop through this formula, the code base grew rather big without noticing this issue and changing to a different convention for representing lack of labels would be very cumbersome.

**Comment by SsnL** [comment]
@jatentaki For that specific workaround, you'd need to write out the backward formula for `f` and `g`.

**Comment by jatentaki** [comment]
@SsnL this sounded scary and unmaintainable, but I think I found an automated workaround. Could this be useful for upstreaming? In any case, I'll leave it in case someone else encounters this issue.

```python
class NaNWhere(torch.autograd.Function):
    '''                                 
    An adaptation of torch.where to situations like
    output = torch.where(torch.isfinite(f1(x)), f1(x), f2(x))
    where the point of torch.where is to mask out invalid application
    of f1 to some elements of x and replace them with fallback values
    provided by f2(x).                                                
    '''                
    @staticmethod
    def forward(ctx, x, f1, f2, mask_fn=torch.isfinite):
        x_1 = x.detach().clone().requires_grad_(True)   
        x_2 = x.detach().clone().requires_grad_(True)
                                                     
        with torch.enable_grad():
            y_1 = f1(x_1)        
            y_2 = f2(x_2)
                         
        mask = mask_fn(y_1)
                           
        ctx.save_for_backward(mask)
        ctx.x_1 = x_1              
        ctx.x_2 = x_2
        ctx.y_1 = y_1 
        ctx.y_2 = y_2
                     
        return torch.where(mask, y_1, y_2)
                                          
    @staticmethod
    def backward(ctx, gout):
        mask, = ctx.saved_tensors
                                 
        torch.autograd.backward([ctx.y_1, ctx.y_2], [gout, gout])
        gin = torch.where(mask, ctx.x_1.grad, ctx.x_2.grad)      
                                                           
        return gin, None, None
                               
nan_where = NaNWhere.apply
```

and in context of a testing script

```python
import torch

def base_where_wrapper(x, f1, f2):
    '''
    This will lead to NaN gradients.
    '''
    y1 = f1(x)
    y2 = f2(x)

    return torch.where(torch.isfinite(y1), y1, y2)

class NaNWhere(torch.autograd.Function):
    '''
    An adaptation of torch.where to situations like
    output = torch.where(torch.isfinite(f1(x)), f1(x), f2(x))
    where the point of torch.where is to mask out invalid application
    of f1 to some elements of x and replace them with fallback values
    provided by f2(x).
    '''
    @staticmethod
    def forward(ctx, x, f1, f2, mask_fn=torch.isfinite):
        x_1 = x.detach().clone().requires_grad_(True)
        x_2 = x.detach().clone().requires_grad_(True)

        with torch.enable_grad():
            y_1 = f1(x_1)
            y_2 = f2(x_2)

        mask = mask_fn(y_1)

        ctx.save_for_backward(mask)
        ctx.x_1 = x_1
        ctx.x_2 = x_2
        ctx.y_1 = y_1
        ctx.y_2 = y_2

        return torch.where(mask, y_1, y_2)

    @staticmethod
    def backward(ctx, gout):
        mask, = ctx.saved_tensors

        torch.autograd.backward([ctx.y_1, ctx.y_2], [gout, gout])
        gin = torch.where(mask, ctx.x_1.grad, ctx.x_2.grad)

        return gin, None, None

nan_where = NaNWhere.apply

def test_case(where_impl):
    x = torch.linspace(-5, 5, 10, requires_grad=True) # includes negative numbers
    
    def f1(x):
        return x.sqrt() # sqrt will output NaN for negative numbers

    def f2(x):
        return (-x).sqrt()

    selected = where_impl(x, f1, f2)
    
    selected.sum().backward()

    return torch.isfinite(x.grad).all()

if __name__ == '__main__':
    print(test_case(base_where_wrapper))
    print(test_case(nan_where))
```

**Comment by SsnL** [comment]
@jatentaki Nice! Thanks for sharing!

**Comment by george-qi** [comment]
Hi all, just wanted to surface some attention for MaskedTensor, a tensor extension library that we just launched the prototype for along PyTorch 1.11. It should help solve many existing problems related to masked semantics, distinguishing between 0 and undefined gradients, etc. 

e.g. for this example:

```
a = masked_tensor(torch.randn(()), torch.tensor(True), requires_grad=True)
b = torch.tensor(False)
c = torch.ones(())

print(torch.where(b, a/0, c))
masked_tensor(  1.0000, True)
print(torch.autograd.grad(torch.where(b, a/0, c), a))
(masked_tensor(--, False),)
```

For more details, you can check out the [website](pytorch.org/maskedtensor) or [Github](github.com/pytorch/maskedtensor). We're actively looking for new users, so if you're interested in using it or have any suggestions on things we should build out, please reach out to me or @cpuhrsch

**Comment by vadimkantorov** [comment]
regardless of the "proper" solution, I think it would still be nice to have in addition F.replace / F.sanitize function / pseudo-function to replace specified special values in the map (e.g. inf/nan -> 0) in outputs and/or gradients.

**Comment by soulitzer** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/4132
